Adds support for ctrl+tab and ctrl+shift+tab

snaiperskaya96 opened the pull request.

sc0w commented:
- the indentation is erroneous

- I see the build warning:

```
pluma-window.c:329:54: warning: passing argument 2 of ‘_pluma_cmd_documents_previous_document’ from incompatible pointer type [-Wincompatible-pointer-types]
         _pluma_cmd_documents_previous_document(NULL, window);
```

- The PR doesn't work here, ctrl+tab changes the focus to the toolbar

sc0w commented:
I noticed we have two PRs that fixes the same issue

The PR https://github.com/mate-desktop/pluma/pull/242 was opened before

snaiperskaya96 commented:
>the indentation is erroneous

Visual studio code didn't show any difference :roll_eyes: Fixed

>I see the build warning:
pluma-window.c:329:54: warning: passing argument 2 of ‘_pluma_cmd_documents_previous_document’ from incompatible pointer type [-Wincompatible-pointer-types]
         _pluma_cmd_documents_previous_document(NULL, window);

Fixed
>The PR doesn't work here, ctrl+tab changes the focus to the toolbar

Now that was a weird one. When i first made this PR i was trying it on a virtual machine on osx, using the standard Mac keyboard and it seemed like it worked well. Made a change, should work correctly now.

>I noticed we have two PRs that fixes the same issue

Well if it works just merge that one then :smile:

sc0w commented:
ok, the warnings are gone, and it works

but, how to access to the toolbar only with keyboard? That feature was removed with this PR

snaiperskaya96 commented:
@sc0w Mmh i would use Alt + F(ile), Alt + E(dit) and so on...
I modified the PR so the event propagation is not blocked anymore by the tab switching. Is it okay like this?

sc0w commented:
@snaiperskaya96 No, if I am over a item in the toolbar and I press ctrl+tab, I go to the next tab, and this is not the expected for me, it must go to the next item in the toolbar

I suggest these changes, I did the commit https://github.com/sc0w/pluma/commit/646c6f7cb2740bcf34eca4e9ea6a35bbcce27346 over https://github.com/sc0w/pluma/tree/ctrl%2Btab

That commit can used cleanly here if you like

snaiperskaya96 commented:
I merged that commit here 👍

sc0w commented:
ok, but I suggest to apply https://github.com/sc0w/pluma/commit/646c6f7cb2740bcf34eca4e9ea6a35bbcce27346 instead https://github.com/mate-desktop/pluma/pull/243/commits/b9b9fd70577c527abaf469d689a6713833bae189 , I am the author of that commit :)

( I used **{** in another places to keep the original appareance of pluma-window.c )

snaiperskaya96 commented:
@sc0w Okay, pushed YOUR changes 😛

lukefromdc commented:
This works fine, allows navigating multiple tabs and the toolbar entirely from the keyboard

raveit65 commented:
Ok , this works.
but how can i select the statusbar?
Not related to your PR, but this seems to broken in master.

sc0w commented:
I can with ctrl+tab , but with the PR only works on the latest tab, so, ctrl+tab has conflicts with the currently behavior

raveit65 commented:
Ok, without the PR and after switching to TOK theme i can select pluma statusbar with the keyboard.
Btw. i see that dropboxes are selected.

With PR this works only if the last *tab* is selected.
If let's say the first *tab* is selected and the *document view* is active, than using crtl+tab switch the *tab* instead of switching to statusbar.
So we have a little regression.
Can this be improved please?

raveit65 commented:
@snaiperskaya96 
Keyboard-navigation is used by people with visual impairments, so it is important that everything works as expected and nothing will confused them.

snaiperskaya96 commented:
How do you select the statusbar usually?

raveit65 commented:
If the **document view** is active than you can switch crtl+tab to the status bar.

sc0w commented:
It would be nice a gsettings key to enable ctrl+tab to switch tabs in pluma, it has conflicts with the current behavior

snaiperskaya96 commented:
>If the document view is active than you can switch with crtl+tab to the status bar.

@raveit65 So correct me if i'm wrong. The problem now is that if we ctrl+shift+tab when the first tab is active we go to the status bar and *that isn't right*.

raveit65 commented:
> So correct me if i'm wrong. The problem now is that if we ctrl+shift+tab when the first tab is active we go to the status bar and that isn't right.

No, the switching order in general is toolbar --> close-button sidebar --> sidebar--> bottom notebook tabs of the sidebar --> notebook tabs --> document view --> statusbar.

Try this with Traditional OK theme, which will easily show you what is active. 

> Edit2:
> What is the "standard" way to focus the status bar?

For keyboard navigation we use the *tab* key in general.
But if the **document view** is active than you can't use  **tab** key alone, because this will write a tab in the document itself.
So, here you need to use **crtl+tab** to switch from **document view** to the statusbar.

The problem with your PR is that swiching from **document view** to the statusbar with  **crtl+tab**, only works if the last notebook tab is selected.

I hope this makes it clearer :-)

snaiperskaya96 commented:
So I guess we have to add the change that ZenWalker suggested?

>It would be nice a gsettings key to enable ctrl+tab to switch tabs in
pluma, it has conflicts with the current behavior

2017-08-29 11:11 GMT+01:00 raveit65 <notifications@github.com>:

> So correct me if i'm wrong. The problem now is that if we ctrl+shift+tab
> when the first tab is active we go to the status bar and that isn't right.
>
> No, the switching order in general is toolbar --> close-button sidebar -->
> sidebar--> bottom notebook tabs of the sidebar --> notebook tabs -->
> document view --> statusbar.
>
> Try this with Traditional OK theme, which will easily show you what is
> active.
>
> Edit2:
> What is the "standard" way to focus the status bar?
>
> For keyboard navigation we use the *tab* key in general.
> But if the *document view* is active than you can't use *tab* key alone,
> because this will write a tab in the document itself.
> So, here you need to use *crtl+tab* to switch from *document view* to the
> statusbar.
>
> The problem with your PR is that swiching from *document view* to the
> statusbar with *crtl+tab*, only works if the last notebook tab is
> selected.
>
> I hope this makes it clearer :-)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/mate-desktop/pluma/pull/243#issuecomment-325619119>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFCOMVzFvCftsNC6bH0C-2uA-HB6m10uks5sc-O7gaJpZM4OwrgC>
> .
>

raveit65 commented:
OMG, last commit makes i much worse, now i can't leave the **document view** anymore only with using the keyboard.
You need to understand the general using of *crtl* and *crtl+tab* for keyboard navigation in an application.
Another usage of crtl+tab is to switch to other buttons in the toolbar here in pluma or caja, etc., if the toolbar is active.
This works out of box with your PR, here we have no conflict with using crtl+ tab in general or for notebook tabs.
So this should work too if the **document view** is active.

> So I guess we have to add the change that ZenWalker suggested?

Not sure if we should enable something with gesettings which does not work correctly.

Sorry, english isn't my native language, i can't explain it better.

snaiperskaya96 commented:
Ignore that, i made it before you did explain how it should be :p
I think i've understood what you meant now, will have another try later.
Thanks for your time btw :)

2017-08-29 11:52 GMT+01:00 raveit65 <notifications@github.com>:

> OMG, last commit makes i much worse, now i can't leave the *document view*
> anymore only with using the keyboard.
> You need to understand the general using of *crtl* and *crtl+tab* for
> keyboard navigation in an application.
> Another usage of crtl+tab is to switch to other buttons in the toolbar
> here in pluma or caja, etc., if the toolbar is active.
> This works out of box with your PR, here we have no conflict with using
> crtl+ tab in general or for notebook tabs.
> So this should work too if the *document view* is active.
>
> So I guess we have to add the change that ZenWalker suggested?
>
> Not sure if we should enable something with gesettings which does not work
> correctly.
>
> Sorry, english isn't my native language, i can't explain it better.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/mate-desktop/pluma/pull/243#issuecomment-325627992>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFCOMVSFAkdIXmWaRByl1vZTL3924cFCks5sc-1-gaJpZM4OwrgC>
> .
>

alexarnaud commented:
To be honest i don't understand why ctrl+tab was used to focus the "status bar". The proper keyboard shortcut should be F6 for that stuff. Moving between the items on the bar should be with tab.

Best regards.

raveit65 commented:
> The proper keyboard shortcut should be F6 for that stuff
Why?
And where is that written?
Again, the general key for tapping through an application is the tab key, for all applications.
And in **document view** the crtl do mask tab key, to avoid to  write a tab in the document itself.
In other applications like eom there is no need to use crtl+tab to select the toolbar, because the tab key can't write in a image as it isn't a text.
The is a very general concept and should not be broken.
Please make this PR not more complicated as needed :-)

alexarnaud commented:
Le 29/08/2017 à 17:23, raveit65 a écrit :
>     The proper keyboard shortcut should be F6 for that stuff
>     Why?
>     And where is that written?

It is documented on the Wikipedia page of keyboard shortcuts, see 
https://en.wikipedia.org/wiki/Table_of_keyboard_shortcuts, search for 
the line "Give focus to next/previous pane"

Also on MSDN, e.g. 
https://msdn.microsoft.com/en-us/library/hh179479(v=nav.90).aspx

It is the behavior on LibreOffice GTK2/GTK3 and maybe all LibreOffice 
front-end, see https://help.libreoffice.org/Common/General_Shortcut_Keys_in

Maybe people are more experienced than me on keyboard shortcuts usage 
but I haver never heard about ctrL+tab except for leaving texts areas 
which catch the tab key.

Otherwise the use of ctrl+tab is documented on the GNOME side, e.g. 
https://help.gnome.org/users/gnome-help/stable/keyboard-nav.html.

Also the majority of person from windows knows ctrl+tab for switching 
between tabs and some people know F6 to switch to applications bars.

Best regards.
-- 
Alex ARNAUD
Visual-Impairment Project Manager
Hypra - "Humanizing technology"

raveit65 commented:
> Maybe people are more experienced than me on keyboard shortcuts usage but I haver never heard about ctrL+tab except for leaving texts areas which catch the tab key.

......exactly this is broken.

raveit65 commented:
And if you  want to reorganize the complete keyboard navigation behahavoiur, please open another report and find a person who wants to rewrite the code.
But here we have the problem the crtl-tab is broken to leave the text area, which is the way to navigate to the statusbar.

snaiperskaya96 commented:
Sorry about that but i just want to be sure. You are saying 
>So, here you need to use crtl+tab to switch from document view to the statusbar.

At the moment (ignore the last commit) i use ctrl+tab to go to the next tab as requested by this ticket https://github.com/mate-desktop/pluma/issues/211 . You are saying that ctrl+tab should instead be used to go from the document view to the status bar.
So how should the user switch tab when the document view is focused?

I think we are going out of the ticket scope.

alexarnaud commented:
Le 29/08/2017 à 18:08, raveit65 a écrit :
> And if you want to reorganize the complete keyboard navigation 
> behahavoiur, please open another report and find a person who wants to 
> rewrite the code.
> But here we have the problem the crtl-tab is broken to leave the text 
> area, which is the way to navigate to the statusbar.

Right :). So the GSettings key should be shipped with a checkbox in 
settings to make the new feature user-friendly.

Best regards.
-- 
Alex ARNAUD
Visual-Impairment Project Manager
Hypra - "Humanizing technology"

raveit65 commented:
> At the moment (ignore the last commit) i use ctrl+tab to go to the next tab as requested by this ticket >#211 . You are saying that ctrl+tab should instead be used to go from the document view to the status bar.

No. ctrl+tab is the current behaviour to navigate from textview to statusbar.

>So how should the user switch tab when the document view is focused?

> I think we are going out of the ticket scope.

Sigh, please try keyboard navigation without your PR.
Than you will understand what i am saying.
I am a little bit tired to write it again here, and you never try to navigate from textview to statusbar for yourself.

raveit65 commented:
@snaiperskaya96 
I will try to say it again in simple sentences.

Without your PR crt+tab is used for
- for selecting buttons inside the toolbar (1)
- for switching from textview to statusbar (2)

With your PR
- for switching between tabs of textview (new) (3)
- for selecting buttons inside the toolbar (works well) (1)
- for switching from textview to statusbar (broken) (2)

As (2) is working with your PR , i don't think that a gsettings key is needed to fix the problem.

PS: If this isn't understandable than someone other needs to explain the problem.......

sc0w commented:
clarification: with this PR it is possible to acces to the toolbar/statusbar/sidebar and other items since the commit https://github.com/mate-desktop/pluma/pull/243/commits/ed6cd6e542c812c90b8b3608f46d9e6d969c933a

but it only works in the last tab with [ctrl+tab] or with the first tab with [ctrl+shift+tab]

@raveit65 
> for selecting buttons inside the toolbar (works well) (1)

It doesn't work well, for example the "save" button, with this PR it only can be selected with [ctrl+tab] for the last/first tab, we can't save other tabs with that item only with keyboard

without this PR, for example, with three tabs, we are in the middle, the second tab, it is possible to access to the statusbar or the toolbar without change the tab, directly with [ctrl+tab] and [ctrl+shift+tab]

it is impossible to give two behaviors to that keys, so, honestly, in my opinion, why not give to the user the power to choice?

I think there is only one way: modify the PR to work [ctrl+tab] and [ctrl/shift+tab] **only** to change between tabs, and add gsettings key with this option, but the standard as default (the classic, to change between items)

I think, the user like to change tabs with [ctrl+tab] doesn't expect another behavior with that keys
